File: frontend/src/api/types/complaints.ts
export type ApiComplaint = {
  id: string
  public_reference: string
  ward_id: number
  ward_name: string
  ward_code: string | null
  category: string
  description: string
  location_detail: string | null
  status: string
  cluster_count: number
  source: string
  submitted_at: string
  updated_at: string | null
  reporter_phone: string | null
  department_suggestion: string | null
  assigned_department: string | null
  staff_note: string | null
}

export type ApiComplaintListResponse = {
  total: number
  complaints: ApiComplaint[]
}

export type CreateComplaintPayload = {
  ward_id: number
  category: string
  description: string
  location_detail?: string
}

export type UpdateComplaintPayload = {
  status?: string
  assigned_department?: string
  staff_note?: string
}

File: backend/app/schemas/complaints.py
from pydantic import BaseModel, Field


ALLOWED_CATEGORIES = {
    "water",
    "roads",
    "drainage",
    "electricity",
    "health",
    "sanitation",
    "other",
}

ALLOWED_STATUSES = {
    "submitted",
    "under_review",
    "in_progress",
    "resolved",
}


class ComplaintCreateRequest(BaseModel):
    ward_id: int
    category: str = Field(min_length=1, max_length=100)
    description: str = Field(min_length=1)
    location_detail: str | None = None
    citizen_contact: str | None = None


class ComplaintUpdateRequest(BaseModel):
    status: str | None = None
    assigned_department: str | None = Field(default=None, max_length=100)
    staff_note: str | None = Field(default=None, max_length=2000)


class ComplaintResponse(BaseModel):
    id: str
    public_reference: str
    ward_id: int
    ward_name: str
    ward_code: str | None = None
    category: str
    description: str
    location_detail: str | None
    status: str
    cluster_count: int
    source: str
    submitted_at: str
    updated_at: str | None = None
    reporter_phone: str | None
    department_suggestion: str | None
    assigned_department: str | None = None
    staff_note: str | None = None


class ComplaintListResponse(BaseModel):
    total: int
    complaints: list[ComplaintResponse]
